### ISSUE TITLE
False positive for collapsible_else_if with cfg feature usage

### ISSUE THREAD

**Issue by catenacyber** [rust-lang/rust-clippy#15158]
### Summary

Running `cargo clippy` on suricata rust subdirectory now prints
```
warning: this `else { if .. }` block can be collapsed
    --> src/smb/smb.rs:1794:52
     |
1794 |   ...                   } else {
     |  ______________________________^
1795 | | ...                       SCLogDebug!("SMB1 request seen from server to client");
1796 | | ...                       if let Some(frame) = pdu_frame {
1797 | | ...                           frame.add_event(flow, SMBEvent::RequestToClient as u8);
1798 | | ...                       }
1799 | | ...                   }
     | |_______________________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#collapsible_else_if
     = note: `#[warn(clippy::collapsible_else_if)]` on by default
help: collapse nested if block
     |
1794 ~                                             } else if let Some(frame) = pdu_frame {
1795 +                                                 frame.add_event(flow, SMBEvent::RequestToClient as u8);
1796 +                                             }
     |
```

But we do want the `SCLogDebug!` to print a log in debug mode
`cargo clippy --features debug` does not show this.

### Lint Name

collapsible_else_if

### Reproducer

I tried this code:

```rust
use std::env;

#[cfg(feature = "debug")]
#[macro_export]
macro_rules!print_debug {
    ($($arg:tt)*) => {
        println!($($arg)*);
    }
}

#[cfg(not(feature = "debug"))]
#[macro_export]
macro_rules!print_debug {
    ($($arg:tt)*) => {};
}


fn main() {
    if env::args().len() < 2 {
        println!("Not enough many");
    } else {
        print_debug!("Let's go");
        if env::args().len() > 2 {
            println!("So many");
        }
    }
}
```

I saw this happen:

```
warning: this `else { if .. }` block can be collapsed
  --> src/main.rs:21:12
   |
21 |       } else {
   |  ____________^
22 | |         print_debug!("Let's go");
23 | |         if env::args().len() > 2 {
24 | |             println!("So many");
25 | |         }
26 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#collapsible_else_if
   = note: `#[warn(clippy::collapsible_else_if)]` on by default
help: collapse nested if block
   |
21 ~     } else if env::args().len() > 2 {
22 +         println!("So many");
23 +     }
   |
```

I expected to see this happen:
Nothing printed

### Version

```text
rustc 1.88.0 (6b00bc388 2025-06-23)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: x86_64-apple-darwin
release: 1.88.0
LLVM version: 20.1.5
```

**Comment by profetia** [comment]
This should already be fixed in #14906

**Comment by catenacyber** [comment]
> This should already be fixed in #14906

Cool. So, which clippy version should we use ?

**Comment by profetia** [comment]
You may have to use nightly by now.

**Comment by catenacyber** [comment]
Indeed, no more bug with `clippy 0.1.90 (11ad40bb83 2025-06-28)`

Should I close this ? (Before opening, I had only looked for open issues about collapsible_else_if, not closed ones)

**Comment by samueltardieu** [comment]
Yes, it can be closed if it is fixed in our tree.

**Comment by catenacyber** [comment]
Hello, looks like this slipped in again with clippy 0.1.95 (59807616e1 2026-04-14) :

Running `` on suricata gives
```
error: expected one of `=>`, `@`, `if`, or `|`, found `SCLogDebug`
   --> src/smb/smb1.rs:722:13
    |
721 |         SMB1_COMMAND_NT_CREATE_ANDX
    |                                    - expected one of `=>`, `@`, `if`, or `|`
722 |             SCLogDebug!("SMB1_COMMAND_NT_CREATE_ANDX response {:08x}", r.nt_status);
    |             ^^^^^^^^^^ unexpected token

error: expected one of `=>`, `if`, or `|`, found `SCLogDebug`
   --> src/ike/ikev2.rs:139:25
    |
138 |                     IkeV2PayloadContent::KE(ref kex)
    |                                                     - expected one of `=>`, `if`, or `|`
139 |                         SCLogDebug!("KEX {:?}", kex.dh_group);
    |                         ^^^^^^^^^^ unexpected token

error: aborting due to 2 previous errors

Original diagnostics will follow.

warning: this `if` can be collapsed into the outer `match`
   --> src/smb/smb1.rs:723:13
    |
723 | /             if r.nt_status == SMB_NTSTATUS_SUCCESS {
724 | |                 match parse_smb_create_andx_response_record(&r.data[*andx_offset-SMB1_HEADER_SIZE..]) {
725 | |                     Ok((_, cr)) => {
726 | |                         SCLogDebug!("Create AndX {:?}", cr);
...   |
765 | |                 false
766 | |             }
    | |_____________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
    = note: `#[warn(clippy::collapsible_match)]` on by default
help: collapse nested if block
    |
721 ~         SMB1_COMMAND_NT_CREATE_ANDX
722 |             SCLogDebug!("SMB1_COMMAND_NT_CREATE_ANDX response {:08x}", r.nt_status);
723 ~             if r.nt_status == SMB_NTSTATUS_SUCCESS => {
724 |                 match parse_smb_create_andx_response_record(&r.data[*andx_offset-SMB1_HEADER_SIZE..]) {
...
763 |                 }
764 ~             },
    |

warning: this `if` can be collapsed into the outer `match`
   --> src/ike/ikev2.rs:140:25
    |
140 | /                         if direction == Direction::ToClient {
141 | |                             state.ikev2_container.dh_group = kex.dh_group;
142 | |                         }
    | |_________________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
    |
138 ~                     IkeV2PayloadContent::KE(ref kex)
139 |                         SCLogDebug!("KEX {:?}", kex.dh_group);
140 ~                         if direction == Direction::ToClient => {
141 |                             state.ikev2_container.dh_group = kex.dh_group;
142 ~                         }
    |

```
